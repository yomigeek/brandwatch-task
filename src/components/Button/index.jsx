import React from "react";
import styled from "styled-components";

const CustomButton = styled.button`
  background: ${(props) => (props.disabled ? "#5586A4" : "#0c3e5d")};
  display: inline-block;
  border-radius: 5px;
  padding: 10px 0;
  margin: 15px;
  width 150px;
  border: none;
  color: #fff;
  font-size: 12px;
  cursor: ${(props) => (props.disabled ? "default" : "pointer")};
`;

const Button = ({ title, actionHandler, disabled = false }) => {
  return (
    <CustomButton onClick={actionHandler} disabled={disabled}>
      {title}
    </CustomButton>
  );
};

export default Button;
