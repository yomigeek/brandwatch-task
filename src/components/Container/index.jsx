import styled from 'styled-components';

const Box = styled.div`
  width: 100%;
  background: #fff;
`;

const Cover = styled.div`
//   padding: 20px 30px;
`;

const Container = ({ children }) => {
  return (
    <Cover>
      <Box>{children}</Box>
    </Cover>
  );
};

export default Container;
