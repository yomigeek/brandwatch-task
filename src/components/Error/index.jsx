import React from 'react';
import styled from 'styled-components';

const ErrorBox = styled.div`
  padding: 20px;
  background: #f4cac0;
  color: #f26351;
  border: 2px solid #f4cac0;
`;

const Error = ({ message }) => {
  return <ErrorBox>{message}</ErrorBox>;
};

export default Error;
