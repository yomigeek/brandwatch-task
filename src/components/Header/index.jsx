import styled from "styled-components";

const HeaderBox = styled.header`
  height: 80px;
  width: 100%;
  background: #fff1dc;
  text-align: center;
  color: #fff;
`;

const Header = () => {
  return (
    <HeaderBox>
      <a href="/">
        <img
          src="https://res.cloudinary.com/kugoo/image/upload/v1645483046/upload/logo.png"
          alt="logo"
          height="80px"
          width="80px"
        />
      </a>
    </HeaderBox>
  );
};

export default Header;
