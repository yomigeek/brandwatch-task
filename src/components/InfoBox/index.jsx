import React from "react";
import styled from "styled-components";
import { sentimentColorPicker } from "../../helpers/sentimentColorPicker";

const Box = styled.div`
  padding: 20px;
  background: #fff1dc;
  color: #f26351;
  border: 2px solid #fff1dc;
  font-size: 12px;
  text-align: center;
`;

const Title = styled.div`
  font-weight: bold;
  margin-right: 10px;
  margin-bottom: 20px;
  text-align: center;
  font-size: 13px;
  color: #0e2f5a;
`;

const SentimentText = styled.span`
  color: ${({ sentimentType }) =>
    sentimentType ? sentimentColorPicker(sentimentType) : "#0E2F5A"}; ;
`;

const InfoBox = ({ data = [] }) => {
  return (
    <Box>
      {data?.length < 1 ? (
        <div>
          <br />
          No topic selected <br />
          <br />
          Please click "View" to select a topic
        </div>
      ) : (
        <>
          <Title>Information on topic {`"${data[0].label}"`}</Title>
          <Title>Total Mentions : {data[0].volume}</Title>

          <Title>
            Positive Mentions:{" "}
            <SentimentText sentimentType="positive">
              {data[0].sentiment.positive}
            </SentimentText>
          </Title>
          <Title>
            Neutral Mentions:{" "}
            <SentimentText sentimentType="negative">
              {data[0].sentiment.negative}
            </SentimentText>
          </Title>
          <Title>
            Negative Mentions:{" "}
            <SentimentText sentimentType="neutral">
              {data[0].sentiment.neutral}
            </SentimentText>{" "}
          </Title>
          <br />
        </>
      )}
    </Box>
  );
};

export default InfoBox;
