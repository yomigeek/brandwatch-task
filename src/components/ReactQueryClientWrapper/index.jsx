import { QueryClientProvider, QueryClient } from 'react-query';

const queryClient = new QueryClient();

export const ReactQueryClientWrapper = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
