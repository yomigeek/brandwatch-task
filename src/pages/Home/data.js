export const tableHead = ["label", "sentimentScore", "view"];
