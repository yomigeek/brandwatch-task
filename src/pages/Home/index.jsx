import React, { useState, useEffect } from "react";
import { useQuery } from "react-query";
import styled from "styled-components";
import Header from "../../components/Header";
import Container from "../../components/Container";
import Loader from "../../components/Loader";
import Error from "../../components/Error";
import Table from "../../components/Table";
import TitleBox from "../../components/TitleBox";
import InfoBox from "../../components/InfoBox";
import Button from "../../components/Button";

// data
import { tableHead } from "./data";

// helpers
import { pagination } from "../../helpers/pagination";

// apis
import { getAllTopicsApi } from "./../../api/topics";

const LoaderBox = styled.div`
  text-align: center;
  margin: 120px 200px;
`;

const HomeBox = styled.div`
  display: flex;
  padding: 20px;
`;

const TableBox = styled.div`
  width: 80%;
`;

const DataBox = styled.div`
  display: flex;
  flex-direction: row;
  background: #fff;
  width: 20%;
`;

const CenterBox = styled.div`
  text-align: center;
  padding: 0px;
`;

const Home = () => {
  const [allTopics, setAllTopics] = useState([]);
  const [goFetch, setGoFetch] = useState(false);
  const [currentTopics, setCurrentTopics] = useState([]);
  const [paginationTotal, setPaginationTotal] = useState(10);
  const [singleTopic, setSingleTopic] = useState([]);

  // Triggers the lifecycle that calls the API after 2 seconds
  setTimeout(() => setGoFetch(true), 2000);

  // Mock API Call to fecth all topics using react query

  const {
    isLoading: getAllTopicsLoading,
    error: getAllTopicsError,
    refetch: getAllTopics
  } = useQuery("getAllTopics", () => getAllTopicsApi(), {
    enabled: false,
    retry: false,
    onSuccess: (data) => {
      setAllTopics(data.topics);
    }
  });

  // Lifecycle that is triggerd after 2seconds timeout
  useEffect(() => {
    if (goFetch) {
      getAllTopics();
      clearTimeout();
    }

    // eslint-disable-next-line
  }, [goFetch]);

  // this gets the single topic and set it in state
  const getSingleTopic = (topicId) => {
    const singleTopic = allTopics.filter((item) => item.id === topicId);
    setSingleTopic(singleTopic);
  };

  // This controls the total number of displayed topics to the user on click load
  // more button
  const loadMoreHandler = () => {
    if (paginationTotal < allTopics?.length) {
      setPaginationTotal(paginationTotal + 10);
    }
  };

  // Handles pagination processing using the pagination helper function
  useEffect(() => {
    if (allTopics?.length > 0) {
      const paginatedValues = pagination(allTopics, paginationTotal);
      setCurrentTopics(paginatedValues);
    }

    // eslint-disable-next-line
  }, [allTopics, paginationTotal]);

  return (
    <>
      <Header />
      <Container>
        <HomeBox>
          <TableBox>
            {getAllTopicsLoading || !goFetch ? (
              <LoaderBox>
                <Loader />
              </LoaderBox>
            ) : (
              <>
                {getAllTopicsError ? (
                  <Error message="Failed to fetch topics. Refresh Page again" />
                ) : (
                  <>
                    <TitleBox message="My Topics Challenge" />
                    <Table
                      data={currentTopics}
                      thead={tableHead}
                      viewHandler={getSingleTopic}
                    />
                  </>
                )}
              </>
            )}
          </TableBox>
          <DataBox>{<InfoBox data={singleTopic} />}</DataBox>
        </HomeBox>
      </Container>
      {/* Pagination box */}
      {getAllTopicsLoading || !goFetch ? (
        ""
      ) : (
        <CenterBox>
          <Button title="Load More" actionHandler={loadMoreHandler} />
        </CenterBox>
      )}
    </>
  );
};

export default Home;
