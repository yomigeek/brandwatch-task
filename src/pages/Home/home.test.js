/* eslint-disable no-undef */
import { render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { ReactQueryClientWrapper } from "../../components/ReactQueryClientWrapper";
import Home from "./index";

test("renders page snapshot", () => {
  const history = createMemoryHistory();

  if (typeof window !== "undefined") {
    const matchMediaPolyfill = function matchMediaPolyfill() {
      return {
        matches: false,
        addListener() {},
        removeListener() {}
      };
    };
    window.matchMedia = window.matchMedia || matchMediaPolyfill;
  }
  render(
    <Router history={history}>
      <ReactQueryClientWrapper>
        <Home />
      </ReactQueryClientWrapper>
    </Router>
  );
  const asFragment = render(
    <Router history={history}>
      <ReactQueryClientWrapper>
        <Home />
      </ReactQueryClientWrapper>
    </Router>
  ).asFragment;
  expect(asFragment()).toMatchSnapshot();
});
