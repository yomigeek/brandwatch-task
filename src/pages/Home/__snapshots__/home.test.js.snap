// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders page snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-bdvvtL eEUwUe"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        height="80px"
        src="https://res.cloudinary.com/kugoo/image/upload/v1645483046/upload/logo.png"
        width="80px"
      />
    </a>
  </header>
  <div
    class="sc-dkPtRN kFbBId"
  >
    <div
      class="sc-gsDKAQ lhdjnX"
    >
      <div
        class="sc-jrQzAO gUrQdG"
      >
        <div
          class="sc-kDTinF hHjyFb"
        >
          <div
            class="sc-pVTFL jFaOGg"
          >
            <img
              alt="loader"
              height="70px"
              src="https://res.cloudinary.com/kugoo/image/upload/v1631372885/webimages/loader.svg"
              width="70px"
            />
          </div>
        </div>
        <div
          class="sc-iqseJM kibWwu"
        >
          <div
            class="sc-jRQBWg egqIxt"
          >
            <div>
              <br />
              No topic selected 
              <br />
              <br />
              Please click "View" to select a topic
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
